### PR TITLE
Apply event sourcing to variables

### LINKED ISSUE #6175: Apply event-sourcing to the variable processors
**Description**

* apply the event-sourcing concept to the variable processors
* including restructure the variable state to extract the variable propagation from setting a variable  

Reference:
* [ZEP: Service Task](https://github.com/zeebe-io/enhancements/blob/master/ZEP004-wf-stream-processing.md#the-stream-processor#service-task)

### PR DESCRIPTION
**Pull request by npepinpe**
## Description

This PR applies event-sourcing principles to the variable records. It flips the current state of writing to the RocksDB, which produces a follow up event via the listeners - now we write `Variable.CREATED` and `Variable.UPDATED` events, which are then applied to the state. The whole concept of listeners is removed from the `VariableState`.

The PR focused only on the `Variable` and `VariableDocument` records, and ignored the concept of temporary variables (which have no direct records) and creating/removing scopes, both of which should be handled by different BPMN processors.

It's now expected that setting single variables directly via the state is done only in tests (for setup) or via the `VariableApplier`.

## Related issues

closes #6175 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [x] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

**Review by korthout — APPROVED** [review]
LGTM 👍 Just some minor comments

File: engine/src/main/java/io/zeebe/engine/processing/bpmn/behavior/BpmnStateBehavior.java [hunk 1]
```diff
@@ -222,14 +229,20 @@ public void propagateVariable(final BpmnElementContext context, final DirectBuff
...
-      final long source, final long target, final DeployedWorkflow targetWorkflow) {
-    final var variables = variablesState.getVariablesAsDocument(source);
-    variableDocumentBehavior.mergeDocument(target, targetWorkflow.getKey(), variables);
+  public void copyVariablesToWorkflowInstance(
+      final long sourceScopeKey,
+      final long targetWorkflowInstanceKey,
+      final DeployedWorkflow targetWorkflow) {
+    final var variables = variablesState.getVariablesAsDocument(sourceScopeKey);
+    variableBehavior.mergeDocument(
+        targetWorkflowInstanceKey, targetWorkflow.getKey(), targetWorkflowInstanceKey, variables);
```
[the remark this answers is not in the document]
Comment (npepinpe) [reply engine/src/main/java/io/zeebe/engine/processing/bpmn/behavior/BpmnStateBehavior.java:245]: Good point, it doesn't change anything but the variable state :+1:

[the remark this answers is not in the document]
Comment (npepinpe) [reply engine/src/main/java/io/zeebe/engine/processing/bpmn/behavior/BpmnStateBehavior.java:239]: I thought about it some more, and I'm not so sure anymore. On one hand, there are methods which are just wrappers around the `VariableBehavior`, and you could argue that's where they belong.

On the other hand, the `BpmnBehaviors` often wrap other non `BpmnBehaviors`, which I assume is to decouple the processors from non-bpmn related behaviors, e.g. `CatchEventBehavior` is accessed via the `BpmnEventSubscriptionBehavior`. I think that's also a valid use, and if that was the original intent, we should stick with that.

**Review by korthout — APPROVED** [review]
Thanks for the changes. LGTM 🧃

**Comment by npepinpe** [comment]
Hm, the build failed because the forked VM crashed due to SIGABRT in the DiskSpaceRecoveryClusteredTest.

**Comment by korthout** [comment]
@npepinpe not sure whether that is something on our side, or a CI infra issue with this as symptom. How do you want to continue?

**Comment by npepinpe** [comment]
I don't think it's related to my PR, but we should have a quick look before dismissing it :sweat_smile:

**Comment by npepinpe** [comment]
bors retry

I doubt it has to do with the PR - looking at the core dump, it crashed while taking a snapshot in RocksDB. So while that's worrying, it would most likely be either an environment issue, or how we use RocksDB in general, but not the variable code specifically.

**Pull request merged**